### PR TITLE
Add support for PHP RFC: Flexible Heredoc and Nowdoc Syntaxes

### PR DESCRIPTION
**Pull request by rainx**
This update one jest snapshot file, please check that. let me know if it needs some changes.

### CODE REVIEW

**Comment by rainx** [comment]
> Please add tests

Added

File: src/lexer/strings.js [hunk 1]
```diff
@@ -126,25 +129,100 @@ module.exports = {
...
+    while (/[\t ]/.test(leading_ch)) {
+      if (leading_ch === " ") {
+        indentation_uses_spaces = true;
+      } else if (leading_ch === "\t") {
+        indentation_uses_tabs = true;
+      }
+
+      leading_ch = this._input[offset + indentation];
+      indentation++;
+    }
```
Comment (evilebottnawi) [inline src/lexer/strings.js]: I think we should throw error on invalid indentation like:
```
// Parse error: Invalid body indentation level (expecting an indentation at least 5) in %s on line %d
```

How in rfc

Comment (rainx) [reply src/lexer/strings.js]: Get it. I will added it later

Comment (rainx) [reply src/lexer/strings.js]: > I think we should throw error on invalid indentation like:
> 
> ```
> // Parse error: Invalid body indentation level (expecting an indentation at least 5) in %s on line %d
> ```
> 
> How in rfc

Just added code to detect this kind of error and raise the error via `this.raiseError` , I am not sure if I handle it in the right way. please help check it.

**Comment by evilebottnawi** [comment]
/cc @ichiriac need review too, thanks

for me looks good

File: src/lexer/strings.js [hunk 2]
```diff
@@ -126,25 +129,100 @@ module.exports = {
...
+
+    // skip leading spaces or tabs
+    let indentation_uses_spaces = false;
+    let indentation_uses_tabs = false;
+    // reset heredoc_label structure
+    let indentation = 0;
+    let leading_ch = this._input.substring(offset - 1, offset);
+
+    while (/[\t ]/.test(leading_ch)) {
```
Comment (ichiriac) [inline src/lexer/strings.js]: please, no regex as they perform about 10 times slower than simple char checks `while ( leading_ch == '\t' || leading_ch == ' ') { `

**Comment by ichiriac** [comment]
@evilebottnawi I will finish the review tomorrow

File: src/lexer/strings.js (hunk 2 above)
Comment (rainx) [reply src/lexer/strings.js]: Ok.. just removed this regex in loop

**Comment by rainx** [comment]
> @evilebottnawi I will finish the review tomorrow

Any other updates?

File: src/parser/scalar.js [hunk 3]
```diff
@@ -79,6 +86,65 @@ module.exports = {
...
+      } else {
+        inCoutingState = false;
+      }
+
+      if (inCheckState && leadingWhitespaceCharCount < indentation) {
+        this.raiseError(
+          `Invalid body indentation level (expecting an indentation at least ${indentation})`
+        );
+      } else {
+        inCheckState = false;
```
Comment (ichiriac) [inline src/parser/scalar.js]: Be aware that `raiseError` will not throw an error in `suppressErrors` mode. You shoud avoid else as if the process continues it should fallback in a normal state

File: src/lexer/strings.js (hunk 1 above)
Comment (ichiriac) [reply src/lexer/strings.js]: It's ok, but you should also test the code with `suppressErrors` mode in order to check it's fallback mode

**Comment by ichiriac** [comment]
It's checked and it's all good, thanks for the hard work and sorry for the late reply. I'll make some changes and then release it under a stable version.

Many thanks again for your contribution !

**Pull request merged**